### PR TITLE
upload csv feature data per planning unit [MRXN23-170]

### PR DESCRIPTION
**Pull request by alexeh**
## Substitute this line for a meaningful title for your changes

This PR takes care of US [MRXN23-168](https://vizzuality.atlassian.net/browse/MRXN23-168) and related tasks:

[MRXN23-170](https://vizzuality.atlassian.net/browse/MRXN23-170)
[MRXN23-171](https://vizzuality.atlassian.net/browse/MRXN23-171)

Even tho there is related third task [(MRXN23-179)](https://vizzuality.atlassian.net/browse/MRXN23-179), it was decided to leave this one out of scope.

### Overview

A user can now add features to the project along with the amount of each feature per planning-unit for a specific project.

2 new tables are created, 1 to keep track of the uploaded files, and a second table that will hold the actual data for each upload.

Since this will allegedly be a 2 step process, no geospatial magic is triggered, and only the data held by the uploaded file is stored in the API database, pending some user confirmation over it. Actions to perform after this confirmation are yet to be implemented

Main changes are inside geo-features module, where 2 entities are added to represent the new entities in the data model.
A CSV parser that will perform data shape validations, and a service that will handle the process along with performing more data integrity validations.
A last service is created to handle api events regarding to this process.

2 new tests are added to check the happy flow and one edge case, even tho more edge cases are handled by the code.

If is considered that is not enough, maybe @yulia-bel can add the remaining cases, as it should be an easy job. Can't handle it myself before closing for vacations as one last heavy duty came to my plate

### Testing instructions

A valid CSV can be uploaded to `/projects/:projectId/features/csv`
A new row in `feature_upload_registry` should be created, and `feature_amounts` table should be populated with as many rows as PUID's are in the CSV, each one holding N features and amounts related to it

Consider checking edge cases beyond what it's tested by automation:
-The project with the given projectId must exist
-Features coming from the CSV can't exist in the DB
-`MAX(csv.puid)` can't be greater than `MAX(projects_pu.puid)` for a given project
-...

### Feature relevant tickets

[MRXN23-170](https://vizzuality.atlassian.net/browse/MRXN23-170)
[MRXN23-179](https://vizzuality.atlassian.net/browse/MRXN23-179)

---

## Checklist before submitting

- [ ] Meaningful commits and code rebased on `develop`.
- [ ] If this PR adds feature that should be tested for regressions when
      deploying to staging/production, please add brief testing instructions
      to the deploy checklist (`docs/deployment-checklist.md`)
- [ ] Update CHANGELOG file

[MRXN23-168]: https://vizzuality.atlassian.net/browse/MRXN23-168?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MRXN23-170]: https://vizzuality.atlassian.net/browse/MRXN23-170?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MRXN23-179]: https://vizzuality.atlassian.net/browse/MRXN23-179?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MRXN23-170]: https://vizzuality.atlassian.net/browse/MRXN23-170?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[MRXN23-171]: https://vizzuality.atlassian.net/browse/MRXN23-171?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: api/apps/api/src/modules/geo-features/import/feature-import.events.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+export class FeatureImportEventsService {
+  eventId!: string;
+
+  submit = () => API_EVENT_KINDS.features__csv__import__submitted__v1__alpha;
+  finish = () => API_EVENT_KINDS.features__csv__import__finished__v1__alpha;
+  fail = () => API_EVENT_KINDS.features__csv__import__failed__v1__alpha;
+
+  constructor(private readonly apiEvents: ApiEventsService) {}
+
+  async registerEvent(event: API_EVENT_KINDS, data: any) {
```
[the remark this answers is not in the document]
Comment (hotzevzl) [reply api/apps/api/src/modules/geo-features/import/feature-import.events.ts:15]: @yulia-bel as mentioned earlier this morning, something similar to what I had in mind here (functionally similar to the current implementation by Alex, but - in my very biased view as the author - slightly more expressive) is this: https://github.com/Vizzuality/mgis/blob/develop/mgis-api/src/modules/data-export/data-export.events.service.ts

[the remark this answers is not in the document]
Comment (hotzevzl) [reply api/apps/api/src/modules/geo-features/import/feature-import.events.ts:15]: though in order to limit the risk of having N-thousand slightly different implementations, I'd likely go for a small set of interfaces (such as `StartedEventEmitter` with just a `emitStartedEvent()` method signature, `FinishedEventEmitter`, `FailedEventEmitter` likewise with a single event emitting method signature, and a `ApiEventEmitter` interface with general methods/props _or_ a suitable decorator to set up the base repository at instantiating time), and then implementing one or more of these interfaces according to what events we do actually emit, which in most cases will be at least the set of created/finished/failed, plus potentially other ones. Happy to mob-design this (but happy to keep all of this madness out of this PR: if we want to converge on such a strategy we can add a comment in this PR in this class for event emitters, and define and implement something stable later - and refactor this).

File: api/apps/api/src/modules/geo-features/import/features-amounts-upload-registry.api.entity.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  CreateDateColumn,
+} from 'typeorm';
+import { UploadedFeatureAmount } from '@marxan-api/modules/geo-features/import/features-amounts-data.api.entity';
+
+@Entity('feature_upload_registry')
+export class FeatureAmountUploadRegistry {
+  @PrimaryGeneratedColumn('uuid')
+  id!: string;
+
+  @Column({ name: 'project_id' })
```
Comment (yulia-bel) [inline api/apps/api/src/modules/geo-features/import/features-amounts-upload-registry.api.entity.ts:15]: add column type

File: api/apps/api/src/utils/acl.utils.ts [hunk 3]
```diff
@@ -149,7 +160,16 @@ export const mapAclDomainToHttpError = (
     | typeof projectNotEditable
     | typeof projectNotVisible
     | typeof tagNotFoundForProject
-    | ImportProjectError,
+    | typeof importedFeatureNameAlreadyExist
+    | typeof unknownPuidsInFeatureAmountCsvUpload
+    | typeof projectNotFound
```
Comment (yulia-bel) [inline api/apps/api/src/utils/acl.utils.ts:165]: remove duplicate

**Review by hotzevzl — APPROVED** [review]
